platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -118,11 +109,29 @@ - (void)validateValue:(id)value
     if (self.URL) {
         NSURL *url = self.URL.mgl_URLByStandardizingScheme;
         source->setURL(url.absoluteString.UTF8String);
-    } else {
+    } else if (self.geoJSONData) {
         NSString *string = [[NSString alloc] initWithData:self.geoJSONData encoding:NSUTF8StringEncoding];
         const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
         source->setGeoJSON(geojson);
         _features = MGLFeaturesFromMBGLFeatures(geojson);
+    } else {
+    
+        NS_MUTABLE_ARRAY_OF(NS_DICTIONARY_OF(NSString *, id) *) *featuresArray = [NSMutableArray array];
+        for (id<MGLFeature> feature in self.features) {
+            [featuresArray addObject:[feature featureDictionary]];
+        }
+        
+        NS_DICTIONARY_OF(NSString *, id) *featureCollection = @{
+            @"type":@"FeatureCollection",
+            @"features":featuresArray};
+        
+        NSError *error;
+        NSData *featuresJSONData = [NSJSONSerialization dataWithJSONObject:featureCollection options:0 error:&error];
```

Comment:
https://github.com/mapbox/mapbox-gl-native/pull/6524#discussion_r81261103 suggested removing the pretty printing. Noting also that this will be refactored away in a future commit where we eventually just set `mbgl::GeoJSON{ featureCollection }` and avoid the trip to create JSON data.